qualtran/bloqs/controlled_state_preparation.py

```diff
@@ -0,0 +1,349 @@
+        is a phase offset for each coefficient that has to be corrected. This offset is half of the
+        turn angle applied, and is added to the phase for each coefficient.
+        """
+        offset_angles = [0] * (2**self.n_qubits)
+        for i in range(self.n_qubits):
+            for j in range(2**i):
+                item_range = 2 ** (self.n_qubits - i)
+                # if the rom has value 0 the formula gives 180, when it should be 0
+                if amplitude_rom_vals[i][j] == 0:
+                    offset = 0
+                else:
+                    offset = np.pi * amplitude_rom_vals[i][j] / (2**self.rot_reg_size)
+                for k in range(item_range * j, item_range * (j + 1)):
+                    offset_angles[k] += offset
+        # if the matrix is the adjoint, the angles have to be undone, thus just load -theta
+        if self.adjoint:
+            angles = [offset - np.angle(c) for c, offset in zip(self.state, offset_angles)]
+        else:
+            angles = [np.angle(c) - offset for c, offset in zip(self.state, offset_angles)]
+        rom_values = [RotationTree.angle2RomValue(a, self.rot_reg_size) for a in angles]
+        return rom_values
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
+    r"""Class that performs an array of controlled rotations $Z^{\theta_i/2}$ for a list of angles
+    $\theta$. It uses phase kickback and thus needs a phase gradient state in order to work. This
+    state must be provided externally for efficiency, as it is unaffected and can thus be reused.
+    Refer to [1], section on arbitrary quantum state preparation on page 3.
+
+    Args:
+        n_selections: number of qubits used for encoding the selection register of the QROM, it
+            must be equal to $\lceil \log_2(l_{rv}) \rceil$, where $l_{rv}$ is the number of angles
+            provided.
+        rot_reg_size: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        rom_values: the tuple of values to be loaded in the rom, which correspond to the angle of
+            each rotation. In order to get the rom value that corresponds to a given angle use
+            RotationTree.angle2RomValue(angle, rot_reg_size).
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    n_selections: int
+    rot_reg_size: int
+    rom_values: Tuple
+
+    @property
+    def signature(self):
+        return Signature.build(
+            prepare_control=1,
+            selection=self.n_selections,
+            qubit=1,
+            rot_reg=self.rot_reg_size,
+            phase_gradient=self.rot_reg_size,
+        )
+
+    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
+        """Parameters:
```

Comment:
I think it's clearer to not use **soqs and just spell out the args.

Reply:
I used `**soqs` because in some cases the `selection` soquet is not used (as when preparing the amplitude of the first qubit). I tried calling the function without it and it complained that a selection soquet must be provided. Is there a cleaner alternative to this?